src/librustc/middle/const_eval.rs

```diff
@@ -941,6 +961,16 @@ pub fn eval_const_expr_with_substs<'tcx, S>(tcx: &ty::ctxt<'tcx>,
               Some(def::DefVariant(enum_def, variant_def, _)) => {
                   (lookup_variant_by_id(tcx, enum_def, variant_def), None)
               }
+              Some(def::DefLocal(id)) => {
+                  debug!("DefLocal({:?}): {:?}", id, fn_args);
+                  if let Some(val) = fn_args.and_then(|args| args.get(&id).cloned()) {
+                      return Ok(val);
+                  } else {
+                      (None, None)
+                  }
+              },
+              Some(def::DefFn(id, _)) => return Ok(Function(id)),
+              // FIXME: implement const methods?
```

Comment:
what does your PR do if it attempts to eval a call to a non `const fn` whose body is more complex than a single expression?

From my initial reading of the PR, it seems like it is assuming that it is always given a `const fn` and thus can assume that the `block.expr.as_ref().unwrap()` is all it needs to look at in the body, in particular it is not (I think) double-checking that the block has no statements...